### ISSUE TITLE
Documentation request: Syncing "app state" and form state

### ISSUE THREAD

**Issue by magnattic** [MrWolfZ/ngrx-forms#140]
I realize that there are probably a hundred ways that you can use ngrx-forms in your application, but it would be nice to have some general guidelines or best practices on how to combine the form state with the rest of the ngrx state.

For me personally, this was the biggest obstacle while adopting it. I understood how ngrx-forms itself works and how ngrx works, but I had no grasp on how to combine the two worlds properly. The example apps don't really help here, because they solely focus on the forms state.

From what I gather, you usually keep the "normal application state" separate from the "form state" and synchronize them when necessary, for example when you click submit or when a value turns valid or ... (depending on your requirements).

The "forms state" only holds the information that's relevant for the form, while the "application state" might have slightly different information.

Example:

"Forms state" holds the voucher code that a user inputs as well as the information if the form value is valid, if there are validation errors etc.
"Application state" holds the final voucher object including price information etc.

When I started out with ngrx forms, I had no real clue how to get the forms state over into the application state. Do I use an effect? Do I do it in the reducer? Do I listen to SET_VALUE or do I have to emit events from my components?

I think I figured a rough approach now, but for newcomers to the library some ideas or best practices (examples?) would be really helpful!

**Comment by MrWolfZ** [comment]
Thank you for this feedback. As you have pointed out there are many ways to integrate the form state into your own state tree. Since the "correct" way (if such a thing exists) depends completely on your own app and requirements, it is difficult to provide guidelines for this. However, what is possible is to add a section to the docs which outlines the various options of organizing the state. Not sure if I will do this soon, but I would be happy to accept contributions.

**Comment by magnattic** [comment]
I'm still learning and expanding my ngrx-forms application. When I have a better grasp and some different use cases, I could propose some code snippets for the documentation.

**Comment by vvolodin** [comment]
Any update on this, guys? I'm really struggling to implement a seemingly easy thing:
I have an object in my root store and on one of the pages I'm showing a settings form using ngrx-forms. I'm initializing the form values from the root store object and the object may change at any time so I'm updating the form based on changes to this object. What I would like to do - is listen to any form changes and apply them to the root store object. I just can't figure out how to do it without ending up with an endless loop of actions.

**Comment by magnattic** [comment]
@vvolodin Why do you have a loop? The ngrx-forms actions (like SetValueAction) only get fired if the user changes a value, not when you set it programmatically.

Anyway, what we do right now to get changes from the form to the app state is something like this (very simplified example):
```
@Effect() updateConsumption$ = this.store.pipe(
    select(state => state.productPage.formState.controls.consumption),
    filter(consumptionControl => consumptionControl.isValid),
    map(consumptionControl => consumptionChanged({ consumption: consumptionControl.value })
)
```

If you need more specific help, maybe open a StackOverflow question?

@MrWolfZ Maybe we could get a gitter channel going for this project? (like the [ngrx](https://gitter.im/ngrx/platform) one?) So we could help each other out with simple questions.

**Comment by vvolodin** [comment]
I'm running into a loop because updating a store object after the form state has changed causes form state update in my code which results in a new store object update - hence a loop. The problem is precisely that I can't filter out form state updates from user input as opposed to all updates including my code.
Seems like I can create an @Effect for SET_VALUE ngrx-forms action but then I would need to filter the controlId manually which seems dirty...

**Comment by magnattic** [comment]
> I'm running into a loop because updating a store object after the form state has changed causes form state update in my code which results in a new store object update - hence a loop.
There should not be an action from ngrx-forms unless the user (not the code) does something in the form. So if you just update the form state in your store, it should not trigger an action. If it does, that's where your problem lies.

So in the very worst case, the longest chain of events should look like this:
user edits form -> ngrx-forms action -> form state update -> effect -> app state update -> effect -> form state update -> end

If you need more detailed help, maybe post some code on Stackoverflow and I will be happy to take a look at it. :)

**Comment by magnattic** [comment]
> I'm running into a loop because updating a store object after the form state has changed causes form state update in my code which results in a new store object update - hence a loop.

It's really hard to tell more without looking at your code. Shouldn't the second store update be the same as the first? This shouldn't trigger anything if your are using `store.select()`. Maybe consider using `distinctUntilChanged()`

**Comment by vvolodin** [comment]
First of all, thanks a lot for your replies! Yes, I understand it's hard to figure out what's going on in my code. And yes, `distinctUntilChanged` would work if I was willing to handle every control in its own effect. But I'm looking at the whole FormGroupState value - so if anything updates in the form I'm re-creating my root store object.

I've come up with this `@Effect` as a workaround:

```
@Effect()
    updateStylingForm$: Observable<Action> = this.actions$.pipe(
        ofType<SetValueAction<any>>('ngrx/forms/SET_VALUE'),
        filter(action => action.controlId.startsWith(fromSettings.STYLING_FORM_ID)),
        tap(x => console.log('effect', x)),
        withLatestFrom(this.store.select(fromSettings.getStylingFormState)),
        map(([action, formState]) => {
            const form = formState.value;
            //this.updateMarker = String(Math.random());
            return new PatchRootObject({newFormData: form});
        })
    );
```
It works really well but feels very dirty - I'm using a hardcoded action type string (I'm a bit surprised it's not a constant in the library), I'm filtering by my FORM_ID (this might not be too bad though).

**Comment by MrWolfZ** [comment]
@vvolodin Using an effect is a perfectly fine way to go and what you have in your latest post is also what I would have come up with (except that ngrx-forms does in fact export the action types, just not as standalone constants but as static properties on the actions, i.e. `ofType<SetValueAction<any>>(SetValueAction.TYPE)`.

Instead of using an effect however, you can actually use a "listener meta reducer" to update your root store object:

```ts
export function rootObjectFormSyncMetaReducer(
  reducer: ActionReducer<RootState>
): ActionReducer<RootState> {
  return function(state, action) {
    let updatedState = reducer(state, action);

    if (action.type === SetValueAction.TYPE) {
      const setValueAction = action as SetValueAction<any>;
      if (setValueAction.controlId.startsWith(fromSettings.STYLING_FORM_ID)) {
        const newFormData = fromSettings.getStylingFormState(updatedState).value;
        updatedState = reducer(updatedState, new PatchRootObject({ newFormData }));
      }
    }

    // perform sync in other direction as necessary

    return updatedState;
  };
}
```

This has the advantage that the store update becomes atomic and will prevent any kind of loop.

@magnattic This library is small enough that I don't mind having these kinds of discussion here in issues.

**Comment by BovineEnthusiast** [comment]
@MrWolfZ I've read this issue and am not 100% sure if this is the right place to mention this, so let me know if I should create a new issue:

I had a lot of trouble recently because I didn't realize that `onNgrxForms` only checks the top level objects. In my case, we organize the state like this:

```ts
data: {
  formState: {...},
  ...
},
ui: {
...
}
```

I'm not sure if this is something that should be mentioned in the docs or if I should open a new issue to make `onNgrxForms` more generic. Maybe an `onNgrxForm` that would take a locator function similar to `wrapReducerWithFormStateUpdate`?

**Comment by kylebeee** [comment]
@BovineEnthusiast did you solve that at all? I'm just getting started looking at using ngrx-forms in my app and the docs all seem to assume that a root reducer exists.

**Comment by BovineEnthusiast** [comment]
> @BovineEnthusiast did you solve that at all? I'm just getting started looking at using ngrx-forms in my app and the docs seem to assume that a root reducer exists.

We've refactored our design a lot since I've posted and now use Nrwl Nx. Thus, we no longer have the `data` and `ui` sublevels in all our states. Meaning we simply keep any form related state at the root level of any particular state.

If you really need to keep forms in a substate, you can probably implement your own `onNgrxForm` that will take a selector. 

Good luck!

**Comment by kylebeee** [comment]
@BovineEnthusiast Thanks! That's exactly what i ended up doing, i also created a pull request #189.

**Comment by MrWolfZ** [comment]
I've released version [6.3.0](https://github.com/MrWolfZ/ngrx-forms/releases/tag/v6.3.0) which allows `onNgrxForms` to work on form states directly. See my comments in #189 for details on why deeply nested form states as outlined by @BovineEnthusiast are tricky.
